File: leetcode_python/474.py
# 474. 一和零
# https://leetcode-cn.com/problems/ones-and-zeroes/
from typing import List
class Solution:
    # 深搜会爆
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        Len = len(strs)
        dp = [[[0 for _ in range(n + 1)] for _ in  range(m + 1)] for _ in range(Len + 1)]
        for k in range(1, Len + 1):
            cnt0 = strs[k-1].count('0')
            cnt1 = strs[k-1].count('1')
            for i in range(m + 1):
                for j in range(n + 1):
                    dp[k][i][j] = dp[k-1][i][j]             #继承
                    if i - cnt0 >= 0 and j - cnt1 >= 0:     #可更新则更新
                        dp[k][i][j] = max(dp[k][i][j], dp[k-1][i-cnt0][j-cnt1] + 1)
            
        return dp[Len][m][n]

File: leetcode_c_plus_plus/474.cpp
// 474. 一和零
// https://leetcode-cn.com/problems/ones-and-zeroes/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 474.cpp
@Time : 2022/01/02 20:08:51
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Medium
**/
class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        int length = strs.size();
        vector<vector<vector<int>>> dp(length + 1, vector<vector<int>>(m + 1, vector<int>(n + 1)));
        vector<pair<int, int>> nums;
        int count_m;
        int count_n;
        for(string str : strs) {
            count_n = 0;
            count_m = 0;
            for(char ch : str){
                if(ch == '0') count_m++;
                else count_n++;
            }
            nums.push_back(make_pair(count_m, count_n));
        }
        for(int i = 1; i <= length; i++) {
            for(int j = 0; j <= m; j++) {
                for(int k = 0; k <= n; k++) {
                    dp[i][j][k] = dp[i - 1][j][k];
                    if(j >= nums[i - 1].first && k >= nums[i - 1].second) {
                        dp[i][j][k] = max(dp[i][j][k], dp[i - 1][j - nums[i -1].first][k - nums[i - 1].second] + 1);
                    }
                }

            }
        }
        return dp[length][m][n];

    }
};